node.go

```diff
@@ -16,40 +17,51 @@ import (
 	"github.com/cosmos/iavl/internal/encoding"
 )
 
+// NodeKey represents a key of node in the DB.
+type NodeKey struct {
+	version int64
+	nonce   int32
+}
```

Comment:
I'm not really seeing the benefit of `NodeKey` struct and abstraction, it seems to have created a big diff for no benefit that I can see.  Did you consider a type alias like `type NodeKey []byte`?   I think this could remove some code.

Reply:
it will increase the node size in the storage (writeBytes).
the real db key of the node is 12 bytes, but the leftNodeKey can be 2 bytes.
`BigEndian` vs `binaary.PutUvarint`

Reply:
I agree with you, it will reduce some encodeing/decoding operations, but the storage size is more significant IMO

Reply:
I'm not following how the use of a struct to organize code relates to encoding?  All the same information is present.

Reply:
because the encoding style is different in the key and value of database, we are using `binary.PutUvarint` in value writeBytes (to reduct the storage size) and BigEndian in the nodekey (to keep sorted).
If we go with []byte in leftNodeKey, we can't use the `node.leftNodeKey` directly as the db key of left node, it will require extra encoding/decoding.

Reply:
That seems like something we could code around, but it's also just a style choice in the first place, so I'll resolve this conversation. :+1: